simuleval/agents/pipeline.py

```diff
@@ -88,3 +89,170 @@ def __repr__(self) -> str:
+                        f"{child}.source_type({child.source_type}) != {parent}.target_type({parent.target_type}"  # noqa F401
+                    )
+                if child in root_instance:
+                    root_instance.remove(child)
+                output_dict[parent].append(child)
+
+        assert len(root_instance) == 1
+        assert len(leaf_instances) > 0
+        return (root_instance[0], leaf_instances, output_dict)
+
+    @property
+    def source_type(self) -> Optional[str]:
+        return self.source_module.source_type
+
+    @property
+    def target_type(self) -> Optional[List[str]]:
+        if self.output_index is not None:
+            return self.target_modules[self.output_index].target_type
+        return "_".join([target.target_type for target in self.target_modules])
+
+    @property
+    def module_list(self) -> List[GenericAgent]:
+        return self.module_dict.keys()
+
+    def build_states(self) -> Dict[GenericAgent, AgentStates]:
+        return {module: module.build_states() for module in self.module_dict.keys()}
+
+    def push_impl(
+        self,
+        module: GenericAgent,
+        segment: Segment,
+        states: Optional[Dict[GenericAgent, AgentStates]],
+    ):
+        # DFS over the tree
+        children = self.module_dict[module]
+        if len(children) == 0:  # leaf node
+            module.push(segment, states[module])
+            return []
+
+        segment = module.pushpop(segment, states[module])
+
+        outputs = []
+        for child in children:
+            outputs += self.push_impl(child, segment, states)
+        return outputs
+
+    def push(
+        self, segment: Segment, states: Optional[Dict[GenericAgent, AgentStates]] = None
+    ) -> None:
+        if states is None:
+            states = {module: None for module in self.module_dict}
+        else:
+            assert len(states) == len(self.module_dict)
+
+        outputs = self.push_impl(self.source_module, segment, states)
+        return outputs
+        # for index, module in enumerate(self.module_dict[:-1]):
+        #     segment = module.pushpop(segment, states[index])
+        # self.module_dict[-1].push(segment, states[-1])
+
+    def pop(self, states: Optional[Dict[GenericAgent, AgentStates]] = None) -> Segment:
+        outputs = []
+        for module in self.target_modules:
+            if states is None:
+                last_states = None
+            else:
+                assert len(states) == len(self.module_dict)
+                last_states = states[module]
+
+            outputs.append(module.pop(last_states))
+        if self.output_index is not None:
+            return outputs[self.output_index]
+        return outputs
+
+    @classmethod
+    def from_args(cls, args):
+        assert len(cls.pipeline) > 0
+        return cls(
+            {
+                module_class.from_args(args): children
+                for module_class, children in cls.pipeline.items()
```

Comment:
I did this to follow the convention of AgentPipeline (which allows users to define the pipeline with the class only, rather than instances of classes, since the args are added by configuration when initializing the agents)